A bunch of V3 stuff

mtsgrd opened the pull request:
draft commit, will be rewritten

estib-vega commented on apps/desktop/src/lib/state/uiState.svelte.ts:
```diff
@@ -0,0 +1,62 @@
+import { reactive, type Reactive } from '@gitbutler/shared/storeUtils';
+import {
+	createEntityAdapter,
+	createSlice,
+	type EntityState,
+	type ThunkDispatch,
+	type UnknownAction
+} from '@reduxjs/toolkit';
+
+type StateKey = 'drawer';
+
+type DrawerPage = 'commit' | 'pr' | 'br' | undefined;
+
+type UiStateVariable = {
+	id: string;
+	value: string | number | boolean | undefined;
+};
+
+export class UiStateService {
+	private state = $state<EntityState<UiStateVariable, string>>(uiStateSlice.getInitialState());
+
+	constructor(
+		reactiveState: Reactive<typeof this.state>,
+		private dispatch: ThunkDispatch<any, any, UnknownAction>
+	) {
+		$effect(() => {
+			console.log(reactiveState.current);
+			this.state = reactiveState.current;
+		});
+	}
+
```
```ts
import { reactive, type Reactive } from '@gitbutler/shared/storeUtils';
import {
	createEntityAdapter,
	createSlice,
	type EntityState,
	type ThunkDispatch,
	type UnknownAction
} from '@reduxjs/toolkit';

type StateKey = 'drawer' | 'drawerWidth' | 'sideBarOpen';

type DrawerPage = 'commit' | 'pr' | 'br' | undefined;

interface UiStateVariableBase {
	id: StateKey;
	value: string | number | boolean | undefined;
};

interface UiStateDrawerPage extends UiStateVariableBase {
	id: 'drawer';
	value: DrawerPage;
};

interface UiStateDrawerWidth  extends UiStateVariableBase {
	id: 'drawerWidth';
	value: number;
};

interface UiStateSideBarOpen  extends UiStateVariableBase {
	id: 'sideBarOpen';
	value: boolean;
};

type UiStateVariable = UiStateDrawerPage | UiStateDrawerWidth | UiStateSideBarOpen;

type UiStateVariableForKey<T extends StateKey> = Extract<UiStateVariable, { id: T }>;

type UiStateValueForKey<T extends StateKey> = UiStateVariableForKey<T>['value'];

export class UiStateService {
	private state = $state<EntityState<UiStateVariable, StateKey>>(uiStateSlice.getInitialState());

	constructor(
		reactiveState: Reactive<typeof this.state>,
		private dispatch: ThunkDispatch<any, any, UnknownAction>
	) {
		$effect(() => {
			console.log(reactiveState.current);
			this.state = reactiveState.current;
		});
	}

	getById<T extends StateKey>(id: T): Reactive<UiStateVariableForKey<T> | undefined> {
		const selected = $derived(selectById(this.state, id));
		return reactive(() => (selected?.id === id ? (selected as UiStateVariableForKey<T>) : undefined));
	}

	update<T extends StateKey>(id: T, value: UiStateValueForKey<T>): void {
		this.dispatch(upsertOne({ id, value } as UiStateVariable));
	}

	getDrawerPage(): Reactive<DrawerPage> {
		const state = $derived(this.getById('drawer'));
		return reactive(() => state.current?.value);
	}
}

export const uiStateVariableAdapter = createEntityAdapter<UiStateVariable, StateKey>({
	selectId: (item) => item.id
});

const { selectById } = uiStateVariableAdapter.getSelectors();

export const uiStateSlice = createSlice({
	name: 'uiState',
	initialState: uiStateVariableAdapter.getInitialState(),
	reducers: {
		upsertOne: uiStateVariableAdapter.upsertOne
	},
	selectors: { selectById }
});

const { upsertOne } = uiStateSlice.actions;
```
```suggestion
import { reactive, type Reactive } from '@gitbutler/shared/storeUtils';
import {
	createEntityAdapter,
	createSlice,
	type EntityState,
	type ThunkDispatch,
	type UnknownAction
} from '@reduxjs/toolkit';

type StateKey = 'drawer';

type DrawerPage = 'commit' | 'pr' | 'br' | undefined;

type UiStateVariable = {
	id: string;
	value: string | number | boolean | undefined;
};

export class UiStateService {
	private state = $state<EntityState<UiStateVariable, string>>(uiStateSlice.getInitialState());

	constructor(
		reactiveState: Reactive<typeof this.state>,
		private dispatch: ThunkDispatch<any, any, UnknownAction>
	) {
		$effect(() => {
			console.log(reactiveState.current);
			this.state = reactiveState.current;
		});
	}

```

estib-vega commented on apps/desktop/src/lib/state/uiState.svelte.ts:
```diff
@@ -0,0 +1,62 @@
+	type UnknownAction
+} from '@reduxjs/toolkit';
+
+type StateKey = 'drawer';
+
+type DrawerPage = 'commit' | 'pr' | 'br' | undefined;
+
+type UiStateVariable = {
+	id: string;
+	value: string | number | boolean | undefined;
+};
+
+export class UiStateService {
+	private state = $state<EntityState<UiStateVariable, string>>(uiStateSlice.getInitialState());
+
+	constructor(
+		reactiveState: Reactive<typeof this.state>,
+		private dispatch: ThunkDispatch<any, any, UnknownAction>
+	) {
+		$effect(() => {
+			console.log(reactiveState.current);
+			this.state = reactiveState.current;
+		});
+	}
+
+	getById(id: StateKey): Reactive<UiStateVariable | undefined> {
+		const selected = $derived(selectById(this.state, id));
+		return reactive(() => selected);
+	}
+
+	update(id: StateKey, value: string | number | boolean | undefined) {
+		this.dispatch(upsertOne({ id, value }));
+	}
+
+	getDrawerPage(): Reactive<DrawerPage> {
+		const state = $derived(this.getById('drawer'));
+		return reactive(() => state.current?.value as DrawerPage);
+	}
+}
```
```suggestion
type StateKey = 'drawer' | 'drawerWidth' | 'sideBarOpen';

type DrawerPage = 'commit' | 'pr' | 'br' | undefined;

interface UiStateVariableBase {
	id: StateKey;
	value: string | number | boolean | undefined;
};

interface UiStateDrawerPage extends UiStateVariableBase {
	id: 'drawer';
	value: DrawerPage;
};

interface UiStateDrawerWidth  extends UiStateVariableBase {
	id: 'drawerWidth';
	value: number;
};

interface UiStateSideBarOpen  extends UiStateVariableBase {
	id: 'sideBarOpen';
	value: boolean;
};

type UiStateVariable = UiStateDrawerPage | UiStateDrawerWidth | UiStateSideBarOpen;

type UiStateVariableForKey<T extends StateKey> = Extract<UiStateVariable, { id: T }>;

type UiStateValueForKey<T extends StateKey> = UiStateVariableForKey<T>['value'];

export class UiStateService {
	private state = $state<EntityState<UiStateVariable, StateKey>>(uiStateSlice.getInitialState());

	constructor(
		reactiveState: Reactive<typeof this.state>,
		private dispatch: ThunkDispatch<any, any, UnknownAction>
	) {
		$effect(() => {
			console.log(reactiveState.current);
			this.state = reactiveState.current;
		});
	}

	getById<T extends StateKey>(id: T): Reactive<UiStateVariableForKey<T> | undefined> {
		const selected = $derived(selectById(this.state, id));
		return reactive(() => (selected?.id === id ? (selected as UiStateVariableForKey<T>) : undefined));
	}

	update<T extends StateKey>(id: T, value: UiStateValueForKey<T>): void {
		this.dispatch(upsertOne({ id, value } as UiStateVariable));
	}

	getDrawerPage(): Reactive<DrawerPage> {
		const state = $derived(this.getById('drawer'));
		return reactive(() => state.current?.value);
	}
}
```

estib-vega commented on apps/desktop/src/lib/state/uiState.svelte.ts:
```diff
@@ -0,0 +1,147 @@
+		drawer: {
+			page: undefined as DrawerPage
+		},
+		commitGoesHere: {
+			parentId: undefined as string | undefined,
+			branchName: undefined as string | undefined
+		}
```
I wonder if this belongs here.
I thing this should about about ui state that is agnostic of any kind of business logic.

Storing commit information sounds to me more like a selection concern that is separate from UI.
What do you think?

estib-vega commented on apps/desktop/src/lib/state/uiState.svelte.ts:
```diff
@@ -0,0 +1,147 @@
+	 * parameter value, with types corresponding to their default values.
+	 */
+	private buildGlobalProps<T extends DefaultConfig>(param: T): GlobalStore<T> {
+		const props = {} as GlobalStore<DefaultConfig>;
```
nit pick: use a declaration constraint instead of a type cast.
```suggestion
		const props: GlobalStore<DefaultConfig> = {};
```

mtsgrd committed:
Remove minimum size for lexical textbox

mtsgrd committed:
Misc css fixes

mtsgrd committed:
Simplify component name

mtsgrd committed:
Fixup selection view

- new background
- use file list component that includes context menu

mtsgrd committed:
Fix type in UpstreamCommit

- createdAt is milliseconds in `number`

mtsgrd committed:
Add a service for UI state

- separates scoped properties from global properties
- builds accessors according to shape of given defaults
- regular redux entity adapter for fine grained reactivity

not covered: hydration

mtsgrd committed:
Lots of V3 changes
